### ISSUE TITLE
Restaurants demo asserts with remote planning

### ISSUE THREAD

**Issue by mmandlis** [PolymerLabs/arcs#2490]
To reproduce:
- run remote planning for 'user'
- run shell with &plannerOnlyConsumer in the browser
- accept "Find restaurants near you" suggestion
The recipe is instantiated, but an error appears in the console
`Error: Multiple slotid-!859659263222556:user-null:9:d2ab5e542d1ab8944cf5173665239289883fdd43 inner slots cannot be provided`

The arc behaves normally, next suggestions appear and can be accepted, but the assert keep firing.

I'm investigating.

**Comment by mmandlis** [comment]
Once the restaurants suggestion is accepted in shell, the arc is deserialized in remote-planner (and planning is performed). 
Then after a short delay the [Restaurant] store ends up containing each restaurant **twice**. 

This is a transient state, and with the second refresh in shell, the page no longer causes an exception, but the arc has **two** VolatileStorages for each inner-arc handle.

I disabled planning altogether: simply reusing the existing suggestions for the null arc, to create a new arc, but no new planning - and this is still happening.

@piotrswigon @shans - does any of this ring a bell in regards to inner arc serialization and storage? I would really appreciate a second pair of eyes to look at this with at some point. Thanks!

**Comment by mmandlis** [comment]
repro: https://github.com/PolymerLabs/arcs/pull/2492

**Comment by piotrswigon** [comment]
We haven't done anything but thinking about design for inner arc serialization. 

The error is coming from `_initInnerSlotContainer`, which is called when inspecting the rendered DOM for `slotid` attributes. This aligns well with your finding that restaurants are duplicated.

Shouldn't the storage layer disallow duplicates of entities in the collection? That was my understanding @mykmartin . There may very likely be a bug somewhere else that is causing creation of these 2 stores though...

I can take a deep dive and get to the bottom of the issue if needed, but maybe @shans will have some quick insight.

**Comment by piotrswigon** [comment]
@mmandlis do you have an arc with a repro, and if you do can you grab its serialization from the storage?

I'm thinking that we may find the volatile stores of the transformation there - which would be a cue. Transformations code does not expect them to be serialized yet.

**Comment by mykmartin** [comment]
The storage layer doesn't detect duplicate entities; it relies on the id provided in the entity object. Using the same id twice with different entity data replaces the existing version (this is known incorrect behaviour that will be fixed when entity mutations are implemented). But it's quite possible to have the same entity data with two different ids in the same collection.

**Comment by piotrswigon** [comment]
Once suggestion has been accepted, Arc has been inflated by both web shell and the planner shell. Both arcs performed searches for restaurants, attempted to clear the `[Restaurant]` handle and insert the entities into it.

One my machine there was ~3 seconds difference of timing for clear followed by store performed by the web and the planner shells. In both cases the storage layer has noted 0 entities to remove from the collection - I assume this is ok with such slight differences of timing. @mykmartin can comment.

The entities from both searches were identical, so probably should get de-duplicated. However it turns out they IDs include session id - which is random, so even though the data was the same, the Entity IDs were not.

Some things to ponder:
1) Should Entity IDs contain a component that is derived from their content, instead of using the Session ID?
2) Should [Restaurant] store be volatile? It used to be #nosync, it no longer is. This distinction however makes for a brittle programming model.
3) Should the system be quicker to note the entities added to the collection by the other device few seconds later? Maybe, but relying on it would also make the programming model brittle.

@shans to comment on (1).

**Comment by mykmartin** [comment]
Seems like a bad idea to have arcs in two different shells modifying the same handle? Especially since they're both completely clearing the collection.. I would have thought a planning arc would work on a clone of the source data, and/or have its own handles to modify.

**Comment by mmandlis** [comment]
@mykmartin When deserialization happens, the arc's active recipe is instantiated all over again, so planning arc is no different than a regular arc. 
In fact, during my investigation i completely disabled planning, so the remote planner was just deserializing the same arc in parallel.

**Comment by mykmartin** [comment]
Ok, so putting aside the planning aspect: it still seems like a problem to have separate arcs actively modifying the same handle for the same plan at the same time. How does that not lead them to trample on each other? In this case it seems they were both clearing an empty collection, so no harm done, but that's not always going to be true.

**Comment by piotrswigon** [comment]
Some comments:

1) Planner should not be starting up the arc. @shans had an idea of performing planning against a speculative arc. This would work with the caveat of having handles/slots etc. bound to wrong ids, so translation back to original arc would need to happen. Cloning gives back the translation map for ids I think, so that shouldn't be too hard.

Alternatively maybe we could perform planning against an acquiescent arc? I don't think Stratategizer takes advantage of the fact that the arc is actually running (with functional particles in the PEC and all that) - nor it should. We could have some intermediate state with an inflated arc that Strategizer can reference, but without PEC and running particles?

2) Even with (1), the problem still holds for multi-user scenario. Entity IDs should not contain the session I think. The proposal would be for the particle to request an ID for the Entity, the request ID would be prepended with the Arc ID, and the Arc would validate the uniqueness or ensure data equality with the previously created Entity of the same ID. If Entity ID does not validate, it is a runtime error.

In this example the RestaurantFind particle would ask for ID `googlemaps://place_id`, which would end up being `arcId:googlemaps://place_id'.

@mykmartin @shans, does this sound reasonable?

3) At some point we need to think deeper about how to structure arcs that are meant for multi user consumption. Which bits are synced, which are not. I don't see how we could do this automatically - as often for bits reflected in the UI this is a subjective design decision. Today you can put #nosync on a handle to get it to not sync - which we could use in this recipe on the `[Restaurant]` handle if we decided that's the right UI in this demo.
